app/models/authentication.rb

```diff
@@ -1,11 +1,81 @@
 #encoding: utf-8
 class Authentication < ActiveRecord::Base
+  before_destroy :can_be_destroyed?, prepend: true
+
   FACEBOOK = "facebook"
   GOOGLE = "google_oauth2"
   TWITTER = "twitter"
-  MEETUP="meetup"
-  LINKEDIN="linkedin"
-  OPENID="openid"
-  PARMA="parma"
+  MEETUP = "meetup"
+  LINKEDIN = "linkedin"
+  OPENID = "openid"
+  PARMA = "parma"
+  TECNOLOGIEDEMOCRATICHE = "tecnologiedemocratiche"
   belongs_to :user, class_name: 'User', foreign_key: :user_id
+
+  UNDETACHABLE_PROVIDERS = [ TECNOLOGIEDEMOCRATICHE ]
+
+
+  def self.oauth_raw_info(oauth_data)

```

Comment:
perchè questi quattro metodi sono definiti dentro ad Authentication?
A mio parere puoi creare una classe a parte per loro quattro, un semplicissimo PORO (Plain Old ruby Object)
che inizializzi con oauth_data e che fornisce i quattro metodi `oauth_raw_info`, `oauth_provider`, `oauth_uid` e `oauth_user_info`.

Quindi dove lo usi puoi fare:
`tua_classe = TuaClasse.new(oauth_data)`

e richiamare comodamente `tua_classe.oauth_provider` senza doverteli salvare in ulteriori variabili.

Puoi farne il caching dentro la classe stessa. Tipo:
```
def oauth_raw_info
 @oauth_raw_info ||= @oauth_data['raw_info'] || # TD
           @oauth_data['extra'].try(:[], 'raw_info') || # Meetup, Twitter, Google, Linkedin, FB 
           @oauth_data['info'] # Parma
end
```